src/condition/shi_detector.cpp

```diff
@@ -34,7 +34,105 @@ struct shi_result {
+        // Since the type check is performed elsewhere, we don't need to check again
+        auto argc = static_cast<std::size_t>(root.nbEntries);
+        if (argc == 0) {
+            return;
+        }
+
+        // Calculate the final resource length
+        std::size_t resource_len = 0;
+        for (std::size_t i = 0; i < argc; ++i) {
+            const auto &child = root.array[i];
+            if (child.type == DDWAF_OBJ_STRING && child.stringValue != nullptr &&
+                child.nbEntries > 0) {
+                // if the string is valid or non-empty, increase the resource
+                // length + 1 for the extra space when relevant
+                resource_len +=
+                    static_cast<std::size_t>(child.nbEntries) + static_cast<std::size_t>(i > 0);
+            }
+        }
+
+        indices.reserve(argc);
+        resource.reserve(resource_len);
+
+        std::size_t index = 0;
+        for (std::size_t i = 0; i < argc; ++i) {
+            const auto &child = root.array[i];
+            if (child.type != DDWAF_OBJ_STRING || child.stringValue == nullptr ||
+                child.nbEntries == 0) {
+                continue;
+            }
+
+            const std::string_view str{
+                child.stringValue, static_cast<std::size_t>(child.nbEntries)};
+
+            indices.emplace_back(index, index + str.size() - 1);
+
+            index += str.size() + 1;
+
+            if (!resource.empty()) {
+                resource.append(" "sv);
+            }
+            resource.append(str);
+        }
+    }
+
+    std::size_t find(std::string_view str, std::size_t start = 0)
+    {
+        while ((start = resource.find(str, start)) != npos) {
+            auto end = start + str.size() - 1;
+
+            // Ensure that both start and end are within the same argument
+            std::size_t low = 0;
+            std::size_t high = indices.size() - 1;
+            while (high >= low) {
+                auto mid = low + (high - low) / 2;
+                auto [arg_start, arg_end] = indices[mid];
+
+                // If the end of the current argument is after the start,
+                // search the remaining arguments to the right of this one
+                if (arg_end < start) {
+                    low = mid + 1;
+                    continue;
+                }
+
+                // If the start of the current argument is before the end,
+                // search the remaining arguments to the left of this one.
+                if (arg_start > end) {
+                    high = mid - 1;
+                    continue;
+                }
+
+                // If the start and end are within the boundaries of this
+                // argument, we have determined that there is no overlap
+                if (arg_start <= start && arg_end >= end) {
+                    return start;
+                }
+
+                // Otherwise, there's overlap and it's not a valid match.
+                break;
```

Comment:
This seems as though it should work but it doesn't...  I thought maybe the check should be `start_pair.first < cur.first` so that false implies `start_pair.first >= cur.first`, but even then it doesn't work because it does a binary search and it needs to check in the other direction (`iter < value`).

This works:

```cpp
            auto end = start + str.size() - 1;
            // Lower bound returns the first element where the condition is false,
            // which must be equivalent to cur < start_pair for the binary search to
            // work as expected.
            auto res = std::lower_bound(indices.begin(), indices.end(), std::pair{start, 0},
                [](const auto &cur, const auto &start_pair) {
                    return cur.second < start_pair.first;
                });

            if (res != indices.end() && res->first <= start && res->second >= end) {
                return start;
            }
            // Otherwise, there's overlap and it's not a valid match.
```

It is somewhat harder to reason though.